libopenage/unit/producer.cpp

```diff
@@ -534,12 +535,53 @@ void BuildingProducer::initialise(Unit *unit, Player &player) {
 		unit->give_ability(std::make_shared<AttackAbility>());
 	}
 
+	// dropsite attribute
+	std::vector<game_resource> accepted_resources = get_accepted_resources();
+	if (accepted_resources.size() != 0) {
+		unit->add_attribute(std::make_shared<Attribute<attr_type::dropsite>>(accepted_resources));
+	}
+
 	// building can train new units and ungarrison
 	unit->give_ability(std::make_shared<SetPointAbility>());
 	unit->give_ability(std::make_shared<TrainAbility>());
 	unit->give_ability(std::make_shared<UngarrisonAbility>());
 }
 
+std::vector<game_resource> BuildingProducer::get_accepted_resources() {
+	//TODO use a more general approach instead of hard coded ids
+
+	auto id_in = [id = this->id()](std::initializer_list<int> ids){
+		return std::any_of(ids.begin(), ids.end(), [=](int n){ return n == id; });
+	};
+
+	if (this->id() == 109) { //Town center
+		return std::vector<game_resource>{game_resource::wood,
+			game_resource::food,
+			game_resource::gold,
+			game_resource::stone,
+			game_resource::fish};
+	}
+
+	if (id_in({584, 585, 586, 587})) { //Mine
+		return std::vector<game_resource>{game_resource::gold,
+			game_resource::stone};
+	}
+
+	if (id_in({68, 129, 130, 131})) { //Mill
+		return std::vector<game_resource>{game_resource::food};
+	}
+
+	if (id_in({562, 563, 564, 565})) { //Lumberjack camp
+		return std::vector<game_resource>{game_resource::wood};
+	}
+
+	if (id_in({45, 47, 51})) { //Docks
+		return std::vector<game_resource>{game_resource::fish};
+	}
```

Comment:
building a multimap of: gather id -> dropsite id in `GameSpec` before constructing the `UnitType` classes, would allow an inverse lookup in the building type constructors